// @aws-sdk/client-rekognition/dist-cjs/models/models_1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamProcessorParameterToDelete = exports.VideoTooLargeException = exports.UnsearchedFaceReason = void 0;
const RekognitionServiceException_1 = require("./RekognitionServiceException");
exports.UnsearchedFaceReason = {
    EXCEEDS_MAX_FACES: "EXCEEDS_MAX_FACES",
    EXTREME_POSE: "EXTREME_POSE",
    FACE_NOT_LARGEST: "FACE_NOT_LARGEST",
    LOW_BRIGHTNESS: "LOW_BRIGHTNESS",
    LOW_CONFIDENCE: "LOW_CONFIDENCE",
    LOW_FACE_QUALITY: "LOW_FACE_QUALITY",
    LOW_SHARPNESS: "LOW_SHARPNESS",
    SMALL_BOUNDING_BOX: "SMALL_BOUNDING_BOX",
};
class VideoTooLargeException extends RekognitionServiceException_1.RekognitionServiceException {
    constructor(opts) {
        super({
            name: "VideoTooLargeException",
            $fault: "client",
            ...opts,
        });
        this.name = "VideoTooLargeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, VideoTooLargeException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
        this.Logref = opts.Logref;
    }
}
exports.VideoTooLargeException = VideoTooLargeException;
exports.StreamProcessorParameterToDelete = {
    ConnectedHomeMinConfidence: "ConnectedHomeMinConfidence",
    RegionsOfInterest: "RegionsOfInterest",
};
